README


This archive contains the distribution Coro-PrioChannel,
version 0.004:

  Priority message queues for Coro

This software is copyright (c) 2012 by Darin McBride.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Coro-PrioChannel

0.004     2013-03-23
        + Add Coro::PrioChannel::Multi
        + Queues now age items to prevent starvation

0.003     2012-04-13
        + Add optional minimum-priority-level to size. Need to figure it out
          for get, too (blocking here is harder, as will be testing).

0.002     2012-04-11
        + Add github metadata
        + Add some better documentation for put (hey, it's all relative)
        + Add test for size method

0.001     2012-04-10
        + Initial revision based nearly completely on Coro::Channel.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "AnyEvent::Loop" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.88"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Priority message queues for Coro",
  "dist_author" => [
    "Darin McBride <dmcbride\@cpan.org>"
  ],
  "dist_name" => "Coro-PrioChannel",
  "dist_version" => "0.004",
  "license" => "perl",
  "module_name" => "Coro::PrioChannel",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "AnyEvent" => 0,
    "Coro" => 0,
    "Coro::Semaphore" => 0,
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Coro/PrioChannel.pm
package Coro::PrioChannel;
{
  $Coro::PrioChannel::VERSION = '0.004';
}
use strict;
use warnings;

# ABSTRACT: Priority message queues for Coro


use Coro qw(:prio);
use Coro::Semaphore ();

use List::Util qw(first sum);
use AnyEvent ();

sub SGET()      { 0 }
sub SPUT()      { 1 }
sub REPRIO()    { 2 }
sub NEXTCHECK() { 3 }
sub DATA()      { 4 }
sub MAX()       { PRIO_MAX - PRIO_MIN + DATA + 1 }


sub new {
   # we cheat, just like Coro::Channel.
   bless [
      (Coro::Semaphore::_alloc 0), # counts data
      (Coro::Semaphore::_alloc +($_[1] || 2_000_000_000) - 1), # counts remaining space
      $_[2], # reprioritization check time
      (defined $_[2] ? (AnyEvent->now + $_[2]) : undef), # last reprioritization check
      [], # initially empty
   ]
}


sub _put {
   my $after = (AnyEvent->now + $_[0]->[REPRIO]) if defined $_[0]->[REPRIO];
   push @{$_[0][DATA + ($_[2]||PRIO_NORMAL) - PRIO_MIN()]}, [$_[1], $after];
}

sub put {
   $_[0]->reprioritize;
   _put @_;
   Coro::Semaphore::up   $_[0][SGET];
   Coro::Semaphore::down $_[0][SPUT];
}


sub get {
   Coro::Semaphore::down $_[0][SGET];
   Coro::Semaphore::up   $_[0][SPUT];

   my $a = first { $_ && scalar @$_ } reverse @{$_[0]}[DATA..MAX];

   ref $a ? shift(@$a)->[0] : undef;
}


sub reprioritize {
    return unless defined $_[0]->[REPRIO];

    my $now = AnyEvent->now;
    return unless $_[0]->[NEXTCHECK] <= $now;

    my $q = $_[0];
    foreach my $pri (PRIO_MIN .. PRIO_HIGH) {
        my $next_pri = $pri + 1;
        my $idx      = DATA + $pri - PRIO_MIN;
        my @keep;

        foreach my $item (@{$q->[$idx]}) {
            if ($item->[1] <= $now) {
                _put $q, $item->[0], $next_pri;
            } else {
                push @keep, $item;
            }
        }

        $q->[$idx] = \@keep;
    }
    
    $q->[NEXTCHECK] = $now + $q->[REPRIO];
    return;
}


sub shutdown {
   Coro::Semaphore::adjust $_[0][SGET], 1_000_000_000;
}


sub size {
    my $min = @_ > 1 ? $_[1] - PRIO_MIN + DATA : DATA;
    sum map { $_ ? scalar @$_ : 0 } @{$_[0]}[$min..MAX];
}


1;

__END__

=pod

=head1 NAME

Coro::PrioChannel - Priority message queues for Coro

=head1 VERSION

version 0.004

=head1 SYNOPSIS

    use Coro::PrioChannel;

    my $q = Coro::PrioChannel->new($maxsize);
    $q->put("xxx"[, $prio]);

    print $q->get;

=head1 DESCRIPTION

A Coro::PrioChannel is exactly like L<Coro::Channel>, but with priorities.
The priorities are the same as for L<Coro> itself.

Unlike Coro::Channel, you do have to load this module directly.

=head1 METHODS

=over 4

=item new

Create a new channel with the given maximum size.  Giving a size of one
defeats the purpose of a priority queue.  Optionally specify the amount of
time spent in the queue before an item's priority is boosted to avoid
starvation.

=item put

Put the given scalar into the queue.  Optionally provide a priority between
L<Coro>::PRIO_MIN and L<Coro>::PRIO_MAX.

=item get

Return the next element from the queue at the highest priority, waiting if
necessary.

TODO: allow an optional parameter to wait for a message of a minimum priority
level (i.e., ignore messages of lower priority).

=item reprioritize

Reprioritizes the queue, boosting the priority of elements that have been in
the queue for longer than the reprioritization parameter (passed to the
constructor). This method is called automatically by put() and should not need
to be called directly in normal circumstances.

=item shutdown

Same as Coro::Channel.

=item size

Same as Coro::Channel.

An optional parameter allows you to specify the minimum priority level
that you want to check the size against, i.e., to ignore messages of
lower priority.  This can be used for example if you're in the middle of
an action and you want to check if there is a higher-priority message to
deal with before resuming the current activity.  This will not block.

=back

=head1 AUTHOR

Darin McBride <dmcbride@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Darin McBride.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Coro/PrioChannel/Multi.pm
package Coro::PrioChannel::Multi;
{
  $Coro::PrioChannel::Multi::VERSION = '0.004';
}
use strict;
use warnings;

# ABSTRACT: Multiple-listener priority message queues for Coro


use Coro::PrioChannel;
use Scalar::Util qw(weaken);


sub LIM()  { 0 }
sub CHAN() { 1 }

sub new
{
    my $class = shift;
    my $limit = shift;
    my $self  = bless [$limit, []], $class;

    $self;
}


sub clean
{
    my $self = shift;
    @{$self->[CHAN]} = grep { defined } @{$self->[CHAN]};

    # when we pull out the refs this way, they're
    # no longer weakened, so re-weaking everything.
    # (easier than using splice to pull undef items out -
    #  if we get too many readers, we'll re-evaluate if this
    #  is slow.)
    weaken($_) for @{$self->[CHAN]};
}


sub number_of_listeners
{
    my $self = shift;
    $self->clean();
    scalar @{$self->[CHAN]};
}

# debugging aid.
sub _status
{
    my $self = shift;
    $self->clean();
    'Channel=size :: ' . join ":", map { $_ . "=" . $self->[CHAN][$_]->size() } 0..$#$self;
}


# create new channel, add it to $self, ensure it's weakened, and return
# the non-weak version.
sub listen
{
    my $self = shift;
    my $channel = Coro::PrioChannel->new($self->[LIM]);
    push @{$self->[CHAN]}, $channel;

    $self->clean();

    $channel;
}


sub put
{
    my $self = shift;

    $self->clean();

    # if we were really multi-threaded, we'd still
    # have to check if $_ was defined, but Coro eliminates
    # that possibility since nothing else really runs between
    # the clean() above and this (we don't cede)
    $_->put(@_) for (@{$self->[CHAN]});
}


1;

__END__

=pod

=head1 NAME

Coro::PrioChannel::Multi - Multiple-listener priority message queues for Coro

=head1 VERSION

version 0.004

=head1 SYNOPSIS

    use Coro::PrioChannel::Multi;

    my $q = Coro::PrioChannel::Multi->new($maxsize);
    $q->put("xxx"[, $prio]);

    my $l = $q->listen();
    print $l->get; # from Coro::PrioChannel

=head1 DESCRIPTION

A Coro::PrioChannel::Multi is exactly like L<Coro::PrioChannel>, but with
the ability to add (and lose) listeners.

Unlike Coro::Channel, you do have to load this module directly.

Each item that is put into the channel will get sent to all listener(s).
However, there is no deep copy here, if the item put in is a reference,
all listeners will receive a reference to the same object, which will allow
each listening thread to modify it before the next thread sees it.  This
could be construed as a feature.

Messages put into the channel before any listener is set up will be lost.
Messages put into the channel before a listener is set up will not be 
resent to that listener, even if the message is still in some other listener's
channel.

=head1 METHODS

=over 4

=item new

Create a new multi-channel with the given maximum size.  Giving a size of one
defeats the purpose of a priority queue.  However, with multiple listeners,
this should ensure that each listener deals with the item before we add
the next item.

=item clean

Clears out any channels that have gone away.  Shouldn't normally be needed
as the object will generally self-clean.

The concept is that a listener may no longer be interested, and has let
its channel listener go out of scope, which will leave a hole in the
list of listeners.  This method simply clears out the holes.

=item number_of_listeners

Returns a count of the number of listeners still attached to this
channel.

=item listen

Set up a new listener for the channel.  Returns a Coro::PrioChannel object
that you issue a C<-E<gt>get> against.  For example:

    my $l = $cpcm->listen();
    while (my $item = $l->get())
    {
        #...
    }

=item put

Pass a message to all (current) listeners.  Optionally provide a priority
between L<Coro>::PRIO_MIN and L<Coro>::PRIO_MAX.

=back

=head1 AUTHOR

Darin McBride <dmcbride@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Darin McBride.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Priority message queues for Coro",
  "AUTHOR" => "Darin McBride <dmcbride\@cpan.org>",
  "BUILD_REQUIRES" => {
    "AnyEvent::Loop" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.88"
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Coro-PrioChannel",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Coro::PrioChannel",
  "PREREQ_PM" => {
    "AnyEvent" => 0,
    "Coro" => 0,
    "Coro::Semaphore" => 0,
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.004",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



